Documentation/design-docs/single-file-staging.md

```diff
@@ -0,0 +1,142 @@
+# CoreCLR Single-file Staging 
+
+Customers have asked for a single-file packaging solution for CoreCLR. This document describes the stages of developing this solution.
+
+## 0. Ideal Solution
+Ideally, we want a single-file solution that:
+
+* Is compatible with all CoreCLR applications
+* Bundles MSIL, R2R, native code and custom (data) files
+* Doesn't require installation or cleanup steps
+* Runs directly from the bundle, without extracting components to disk
+* Reduces publish-size
+* Improves startup cost
+* Works cohesively with debuggers, profilers, Watson dump etc. 
+
+With this feature-set, cost, and timeline tradeoffs in consideration, we propose the following steps to bring-up the feature.
+
+## 1. Self-Extractor
+
+The first stage is to develop a pack and extract tool.
+While this stage is technically simplistic, the interface and tooling will match the final solution, giving partner teams and potential customers a chance to prepare for adopting the technology.
+
+### 1.1 Description
+
+This stage implements:
+
+* A  packaging tool (bundler) that embeds the application and all of its dependencies (essentially the contents of the publish directory) as resources into a host executable. 
+	* This version will not support compression of the bundled assemblies.
+* When the executable runs, it will extract all of those files into a temporary directory, and then run as though the app was published to that temporary directory.
+
+The mechanism will support:
+
+* Reuse: Extract on first-run, reuse on subsequent runs.
+* Upgrade: Each version of the app extracts to a unique location, supporting side-by-side use of multiple versions.
+* Uninstall: Users can identify and delete extracted files when the app is no longer needed.
+* Access control: Processes running with elevated access can extract to admin-only-writable locations.
+
+### 1.2 Scenarios
+
+Target audience:
+
+* Environments requiring maximal compatibility -- need to embed different kinds of files (IL, native, etc) into one, without losing functionality such as debuggability.
+
+Limitations:
+
+* Unsuitable for environments that require that the app does not perform disk-writes at startup
+
+Advantages:
+
+* Low cost of development
+* Bundler tool can be used as-is for most further stages
+* Provides ability to develop test infrastructure and prototypes
+
+## 2. Run from Bundle: MSIL
+
+### 2.1 Description
+
+This stage improves on Stage 1 in that 
+
+* MSIL files bundled into the single-file will load and execute directly from the executable.
+* Native libraries will still need to 
+	* Remain in the publish directory unmerged, or
+	* Extracted to the disk like the previous self-extractor stage
+* Debugging support is unaffected. 
+
+### 2.2 Scenarios
+
+Target audience:
+
+* Framework dependent purely managed apps that are not ready-to-run compiled.
```

Comment:
What does this has to do with purely managed apps or ready-to-run?

If your target audience are purely managed apps, then you do not need to keep any of the support for native libraries.

Reply:
The description of the target audience (in this and other points) should not refer to technical details of the solution. It should describe the type of app or developer who would want to choose it.

Reply:
In each of the 5 stages, the goal is to support all apps, with fewer things spilled into temporary files in later stages. 

The phrase "Target Audience" may be a bit confusing. I was trying to point out the customers who are best served by this stage -- because for such customers they would have realized the full benefit of single-file publishing (that is, further stages have no impact).